lib/graphql/tracing/opentelementry_tracing.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'opentelemetry'
+
+module OpenTelemetry
+  module Instrumentation
+    module GraphQL
+      module Tracers
+        # GraphQLTracer contains the OpenTelemetry tracer implementation compatible with
+        # the GraphQL tracer API
+        class GraphQLTracer < ::GraphQL::Tracing::PlatformTracing
+          self.platform_keys = {
+            'lex' => 'graphql.lex',
+            'parse' => 'graphql.parse',
+            'validate' => 'graphql.validate',
+            'analyze_query' => 'graphql.analyze_query',
+            'analyze_multiplex' => 'graphql.analyze_multiplex',
+            'execute_query' => 'graphql.execute_query',
+            'execute_query_lazy' => 'graphql.execute_query_lazy',
+            'execute_multiplex' => 'graphql.execute_multiplex'
+          }
+
+          def platform_trace(platform_key, key, data)
+            return yield if platform_key.nil?
+
+            tracer.in_span(platform_key, attributes: attributes_for(key, data)) do |span|
+              yield.tap do |response|
+                errors = response[:errors]&.compact&.map { |e| e.to_h }&.to_json if key == 'validate'
+                unless errors.nil?
+                  span.add_event(
+                    'graphql.validation.error',
+                    attributes: {
+                      'message' => errors
+                    }
+                  )
+                end
+              end
+            end
+          end
+
+          def platform_field_key(type, field, context = GraphQL::Query::NullContext)
+            # If global setting is not enabled, return nil such that nothing is traced
+            return unless config[:enable_platform_field]
+
+            # If context setting is not set or enabled, return nil such that nothing is traced
+            ns = context.namespace(:opentelemetry)
+            return unless ns.key?(:enable_platform_field) && ns[:enable_platform_field]
```

Comment:
This is the new functionality. Treat `config[:enable_platform_field]` as a global on/off setting, but if it is on, allow it to be managed to execution context.